packages/repository/src/relations/has-many/has-many-through.inclusion.resolver.ts

```diff
@@ -0,0 +1,135 @@
+ * @param getThroughRepo - through repository getter i.e. where through
+ * instances are
+ * @param getTargetRepo - target repository getter i.e where target instances
+ * are
+ */
+export function createHasManyThroughInclusionResolver<
+  Through extends Entity,
+  ThroughID,
+  ThroughRelations extends object,
+  Target extends Entity,
+  TargetID,
+  TargetRelations extends object
+>(
+  meta: HasManyDefinition,
+  getThroughRepo: Getter<
+    EntityCrudRepository<Through, ThroughID, ThroughRelations>
+  >,
+  getTargetRepo: Getter<
+    EntityCrudRepository<Target, TargetID, TargetRelations>
+  >,
+): InclusionResolver<Entity, Target> {
+  const relationMeta = resolveHasManyMetadata(meta);
+
+  return async function fetchHasManyThroughModels(
+    entities: Entity[],
+    inclusion: Inclusion,
+    options?: Options,
+  ): Promise<((Target & TargetRelations)[] | undefined)[]> {
+    if (!entities.length) return [];
+
+    debug('Fetching target models for entities:', entities);
+    debug('Relation metadata:', relationMeta);
+
+    const sourceKey = relationMeta.keyFrom;
+    const sourceIds = entities.map(e => (e as AnyObject)[sourceKey]);
+    const targetKey = relationMeta.keyTo as StringKeyOf<Target>;
+    if (!relationMeta.through) {
+      throw new Error(
+        `relationMeta.through must be defined on ${relationMeta}`,
+      );
+    }
+    const throughKeyTo = relationMeta.through.keyTo as StringKeyOf<Through>;
+    const throughKeyFrom = relationMeta.through.keyFrom as StringKeyOf<Through>;
+
+    debug('Parameters:', {
+      sourceKey,
+      sourceIds,
+      targetKey,
+      throughKeyTo,
+      throughKeyFrom,
+    });
+
+    debug(
+      'sourceId types',
+      sourceIds.map(i => typeof i),
+    );
+
+    const throughRepo = await getThroughRepo();
+    const targetRepo = await getTargetRepo();
+
+    // find through models
+    const throughFound = await findByForeignKeys(
+      throughRepo,
+      throughKeyFrom,
+      sourceIds,
+      {}, // scope will be applied at the target level
+      options,
+    );
+
+    const throughResult = flattenTargetsOfOneToManyRelation(
+      sourceIds,
+      throughFound,
+      throughKeyFrom,
+    );
+
+    const result = [];
+
+    // convert from through entities to the target entities
+    for (const entityList of throughResult) {
+      if (entityList) {
+        // get target ids from the through entities by foreign key
+        const targetIds = entityList.map(entity => entity[throughKeyTo]);
+
+        // the explicit types and casts are needed
+        const targetEntityList = await findByForeignKeys<
+          Target,
+          TargetRelations,
+          StringKeyOf<Target>
+        >(
+          targetRepo,
+          targetKey,
+          (targetIds as unknown) as [],
+          inclusion.scope as Filter<Target>,
+          options,
+        );
+        result.push(targetEntityList);
+      } else {
+        // no entities found
+        result.push(entityList);
```

Comment:
I am a bit confused here. Are we pushing found through entities to the `result`?

Reply:
So keeping in mind `flattenTargetsOfOneToManyRelation` returns an entity list or undefined (see [here](https://github.com/strongloop/loopback-next/pull/6280/commits/69ccb6faeea9cbe0f44bf907a89a1d20b6dfd5e9#diff-254cedde0a29600de123a1ec89754c8cR36)). The code is: 

```ts
if (entityList) {...}
else {
 // now it's undefined
  result.push(entityList) // equivalent to result.push(undefined)
}
```
Let me know if you want me to change it to `result.push(undefined)` to make it clearer 👍

Reply:
Make sense, thanks!